### ISSUE TITLE
IEEE test cases ieee17, ieee18flushz, ieee19flushz depend on x86_64 behaviour

### ISSUE THREAD

**Issue by pawosm-arm** [flang-compiler/flang#263]
Following ieee19flushz test case failure on AArch64 machine, I wrote simple program (note that ieee18flushz fails similar way on non-x86_64):

```
program ieeehex
  use ieee_arithmetic
  implicit none
  integer :: i
  logical :: luflow, lfsav(5), lfset(5)
  real*8 :: a, rone, ex
  integer*4 :: ix, iy, iz(2)
  equivalence(ex, iz)

  lfset = .false.
  call ieee_get_halting_mode(ieee_all, lfsav)
  call ieee_set_halting_mode(ieee_all, lfset)
  call ieee_get_underflow_mode(luflow)
  call ieee_set_underflow_mode(.false.)
  rone = 1.0d0
  a = rone
  do i = 1, 1040
    ex = a
    ix = iz(2)
    iy = iz(1)
    a = a * 0.5d0
    write (*, *) i, a, ix, iy, iand(ix, z'7fffffff').eq.0, (iand(ix, z'7fffffff').eq.0).and.(iy.eq.0)
  end do
  call ieee_set_underflow_mode(luflow)
  call ieee_set_halting_mode(ieee_all, lfsav)
end program ieeehex
```

On x86_64 (compiled by flang or gfortran) it produces following output (skipped first 1020 lines):

```
         1021   4.4501477170144028E-308      3145728            0  F  F
         1022   2.2250738585072014E-308      2097152            0  F  F
         1023    0.000000000000000           1048576            0  F  F
         1024    0.000000000000000                 0            0  T  T
         1025    0.000000000000000                 0            0  T  T
         1026    0.000000000000000                 0            0  T  T
         1027    0.000000000000000                 0            0  T  T
         1028    0.000000000000000                 0            0  T  T
         1029    0.000000000000000                 0            0  T  T
         1030    0.000000000000000                 0            0  T  T
         1031    0.000000000000000                 0            0  T  T
         1032    0.000000000000000                 0            0  T  T
         1033    0.000000000000000                 0            0  T  T
         1034    0.000000000000000                 0            0  T  T
         1035    0.000000000000000                 0            0  T  T
         1036    0.000000000000000                 0            0  T  T
         1037    0.000000000000000                 0            0  T  T
         1038    0.000000000000000                 0            0  T  T
         1039    0.000000000000000                 0            0  T  T
         1040    0.000000000000000                 0            0  T  T
```

I suspect that this behavior was a basis for following lines found in ieee19flushz test case:

```
  if (j .le. 1023) then
    aa2(j) = rlogb
  else
    aa2(j) = ieee_value(rlogb, ieee_negative_inf)
  end if
```

Unfortunately, on AArch64 machines the code above produces different results, again, compiled flang or gfortran:

```
         1021   4.4501477170144028E-308      3145728            0  F  F
         1022   2.2250738585072014E-308      2097152            0  F  F
         1023   1.1125369292536007E-308      1048576            0  F  F
         1024   5.5626846462680035E-309       524288            0  F  F
         1025   2.7813423231340017E-309       262144            0  F  F
         1026   1.3906711615670009E-309       131072            0  F  F
         1027   6.9533558078350043E-310        65536            0  F  F
         1028   3.4766779039175022E-310        32768            0  F  F
         1029   1.7383389519587511E-310        16384            0  F  F
         1030   8.6916947597937554E-311         8192            0  F  F
         1031   4.3458473798968777E-311         4096            0  F  F
         1032   2.1729236899484389E-311         2048            0  F  F
         1033   1.0864618449742194E-311         1024            0  F  F
         1034   5.4323092248710971E-312          512            0  F  F
         1035   2.7161546124355486E-312          256            0  F  F
         1036   1.3580773062177743E-312          128            0  F  F
         1037   6.7903865310888714E-313           64            0  F  F
         1038   3.3951932655444357E-313           32            0  F  F
         1039   1.6975966327722179E-313           16            0  F  F
         1040   8.4879831638610893E-314            8            0  F  F
```

```
As a result, whole the test case 'fails':

------------------------------------ executing test ieee19flushz
ieee19flushz.out
test number 1024 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1025 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1026 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1027 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1028 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1029 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1030 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1031 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1032 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1033 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1034 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1035 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1036 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1037 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1038 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1039 FAILED. res 0(00000000)  exp -1(ffffffff)
test number 1040 FAILED. res 0(00000000)  exp -1(ffffffff)
1040 tests completed. 1023 tests passed. 17 tests FAILED.
make: `verify' is up to date.
```

**Comment by pawosm-arm** [comment]
The ieee17 test case is also failing on non-x86_64 when checking the 7th result:
```
print *,(sqa .eq. sqa2),(ra2 .eq. sa2),(sqa2 .eq. ieee_next_after(sqa, a))
lres(7) = (sqa .eq. sqa2) .or. (sqa2 .eq. ieee_next_after(sqa, a))
```

This code prints following results:
1. gfortran on x86: F T F (so check on lres(7) fails)
2. flang on x86: T T F (so check on lres(7) succeeds)
3. gfortran on aarch64: F T F (so check on lres(7) fails)
4. flang on aarch64: F T F (so check on lres(7) fails)

**Comment by pawosm-arm** [comment]
ieee18flushz, ieee19flushz: gfortran is wrong on non-x86_64, these two test cases are ok, thing is, all non-x86_64 architectures should have __fenv_fesetzerodenorm() and __fenv_fegetzerodenorm() implemented. On AArch64 _FPU_GETCW/_FPU_SETCW macros from <fpu_control.h> can be used to get/set FZ bit.

**Comment by pawosm-arm** [comment]
a small fix that makes ieee17 pass on AArch64:
```
 #ifdef TARGET_LINUX_ARM
   integer, private, parameter :: FE_TONEAREST  = 0
+#ifdef TARGET_LLVM_ARM64
+  integer, private, parameter :: FE_DOWNWARD   = X'00800000'
+  integer, private, parameter :: FE_UPWARD     = X'00400000'
+#else
   integer, private, parameter :: FE_DOWNWARD   = X'00400000'
   integer, private, parameter :: FE_UPWARD     = X'00800000'
+#endif
```

**Comment by pawosm-arm** [comment]
Shouldn't it be closed? Fixed by already merged PR #427
